test/Docfx.Build.Engine.Tests/TemplateManagerUnitTest.AdditionalTemplates.cs

```diff
@@ -0,0 +1,225 @@
+        {
+            DisableCustomTemplatesDir();
+        }
+    }
+
+    /// <summary>
+    /// Both DotnetToolsMode & `DOCFX_CUSTOM_TEMPLATES_DIR` specified.
+    /// </summary>
+    [Trait("Related", "TemplateProcessor")]
+    [Fact]
+    public void TestAdditionalTemplatesPath_With_DotnetToolsMode_And_CustomTemplatesPath()
+    {
+        // Arrange
+        var templates = new List<string> { "default" };
+        var manager = new TemplateManager(templates, null, null);
+
+        try
+        {
+            CreateEmptyTemplatesDir();
+            EnableDotNetToolMode();
+            EnableCustomTemplatesDir();
+
+            // Act
+            var results = manager.GetTemplateDirectories().ToArray();
+
+            // Assert
+            results.Should().HaveCount(2);
+            results[0].Should().Be(ExpectedDefaultTemplateDirForDotNetTool);
+            results[1].Should().StartWith(ExpectedDefaultTemplatesDirForCustom);
+        }
+        finally
+        {
+            DisableDotNetToolMode();
+            DisableCustomTemplatesDir();
+            RemoveEmptyTemplatesDir();
+        }
+    }
+
+    /// <summary>
+    /// DotnetToolMode is enabled but invalid path.
+    /// </summary>
+    [Trait("Related", "TemplateProcessor")]
+    [Fact]
+    public void TestAdditionalTemplatesPath_With_DotnetToolMode_InvalidPath()
+    {
+        // Arrange
+        var templates = new List<string> { "default" };
+        var manager = new TemplateManager(templates, null, null);
+        var listener = CreateTestLoggerListener();
+        Logger.RegisterListener(listener);
+
+        try
+        {
+            EnableDotNetToolMode();
+            // CreateEmptyTemplatesDir(); // Don't create templates dir.
+
+            // Act
+            using var scope = new LoggerPhaseScope(PhaseNameForFilter);
+            var results = manager.GetTemplateDirectories().ToArray();
+
+            // Assert
+            results.Should().HaveCount(0);
+            listener.Items.Should().HaveCount(1);
+            listener.Items[0].Message.Should().Be($".NET Tools templates directory is not found. Path: {DotnetToolTemplatesDir}");
+        }
+        finally
+        {
+            Logger.UnregisterListener(listener);
+            DisableDotNetToolMode();
+        }
+    }
+
+    /// <summary>
+    /// `DOCFX_CUSTOM_TEMPLATES_DIR` is specified but invalid path.
+    /// </summary>
+    [Trait("Related", "TemplateProcessor")]
+    [Fact]
+    public void TestAdditionalTemplatesPath_With_CustomTemplatesPath_InvalidPath()
+    {
+        // Arrange
+        var templates = new List<string> { "default" };
+        var manager = new TemplateManager(templates, null, null);
+        var listener = CreateTestLoggerListener();
+        Logger.RegisterListener(listener);
+
+        try
+        {
+            const string DummyTemplatesDir = @"Z:\Temp\DummyTemplatesDir";
+            Environment.SetEnvironmentVariable(DOCFX_CUSTOM_TEMPLATES_DIR, DummyTemplatesDir);
```

Comment:
Thanks for your response.
It seems good solution to separate globally used context by using `AsyncLocal`.

I thought If it can provide `EnvironmentContext` per tests.
More tests that are marked with "docfx STA" can be executed in parallel and tests run faster.